### PR TITLE
Set an environment variable for tests to find executables.

### LINKED ISSUE #5758: How to find binary from integration test?
Is there a cleaner way to find a binary executable from an integration test other than looking at `current_exe()` and finding it from there?  Cargo itself uses this logic:

https://github.com/rust-lang/cargo/blob/848250597527659dd54b533c3d35d6505dd60a20/tests/testsuite/cargotest/support/mod.rs#L392-L409

I've seen others use `Command::new("./target/debug/NAME")` which can fail in some cases.

It seems like this should be easier.  Am I missing a better way?  If not, would it make sense to provide some method to more easily find binaries?

### PR DESCRIPTION
**Pull request by ehuss**
This adds the environment variable `CARGO_BIN_EXE_<name>` so that integration tests can find binaries to execute, instead of doing things like inspecting `env::current_exe()`.

The use of uppercase is primarily motivated by Windows whose Rust implementation behaves in a strange way.  It always ascii-upper-cases keys to implement case-insensitive matching (which loses the original case).  Seems less likely to result in confusion?

Closes #5758.

### CODE REVIEW

**Comment by Eh2406** [comment]
I like it. But I think the team should discuss before we merge.

**Comment by alexcrichton** [comment]
@rfcbot fcp merge

Looks like a great feature to me!

**Comment by ehuss** [comment]
TIL that HFS normalizes unicode differently from APFS.

**Comment by joshtriplett** [comment]
@rfcbot concern should-use-same-case-as-Cargo.toml

**Comment by ehuss** [comment]
I'll split the unrelated things out.

As for the uppercase thing, I think it will be extremely difficult to support Windows without it.  For example, the Rust standard library loses case ([here](https://github.com/rust-lang/rust/blob/8a87b945b27b5670ac5ed665bbb0fccc1b88a0a0/src/libstd/sys/windows/process.rs#L38)).  In general, supporting case-sensitive environment variables can be difficult (and also unusual in my experience).  Are there specific concerns about what could happen?  I'm well aware that that there can be difficult edge cases, but I think for the vast majority of cases it should be fine.

**Comment by joshtriplett** [comment]
I appreciate that Windows is not case-sensitive. Unix and Linux, however, are. If there's no better solution available for Windows than uppercasing the variable name, then that should *only* occur on Windows.

Files differing only by case do indeed appear in real projects and real use cases. Furthermore, on a case-sensitive platform, changing the case of a binary name refers to a different binary, and users of such platforms (myself included) will find that behavior surprising, unexpected, and unwelcome.

**Comment by joshtriplett** [comment]
Also, different environments uppercase Unicode differently, and some Microsoft tools and libraries will handle the example you have (ß) differently.

**Comment by alexcrichton** [comment]
I disagree that we should be doing something different only on Windows here, that's generally not what Cargo does and can cause compatibility issues when your project works on Linux and doesn't work on Windows. In general I believe we try to have uniform behavior across all platforms, which in cases like this involves taking the intersection of possible behaivors and making that the API that we expose.

I don't think that we really need to worry all that much about unicode here. While it's of course a minefield if you're constantly knee-deep in unicode we're talking about Cargo projects and binary names, which 99.9% of the time don't even contain anything beyond ASCII. I don't think that an extremely small slice of projects which may have odd upper-casings in unicode should cause us to force *everyone* to deal with Windows-specific behavior.

**Comment by joshtriplett** [comment]
@alexcrichton I don't want everyone to deal with Windows-specific behavior either. Uppercasing on all platforms makes everyone have to deal with case-insensitivity, which is Windows-specific behavior. On a Linux platform, I expect names to be case sensitive, and squelching case is a bug.

Programs distinguished only by case isn't a theoretical issue, it's quite common. As one example, many Linux systems have both `/usr/bin/HEAD` and `/usr/bin/head`, which are two completely different programs. (The former sends an HTTP HEAD request, the latter outputs the first lines of a file.)

What I'm suggesting is this: use the actual case on all platforms, have test programs reference environment variables using the actual case on all platforms, and we can have an internal workaround (that we don't have to expose to the user) if it turns out that that doesn't work on Windows. It sounds like Rust on Windows already uppercases environment variable names, so wouldn't that Just Work? Cargo would set the environment variable by name (which gets internally uppercased on Windows), and rustc would read the environment variable by name (which gets internally uppercased on Windows), so they'd match and write/read the same environment variable.

**Comment by ehuss** [comment]
I'll need to do some tests on Windows to see how it treats unicode.  I know the Rust std lib only handles ascii, but I'm not sure what Windows itself does (I can't find any info online).  I'll investigate and see what issues arise.  In theory the rustc side should also be case-insensitive on windows so it shouldn't matter what case it is, but I suspect it is broken for unicode.

Are you aware of examples of tools with environment variables that use case-sensitive mixed-case variable names?  It seems very unusual to me.

> Programs distinguished only by case isn't a theoretical issue,

I think that's a bit of a stretch, this only matters when the programs are in the same Cargo workspace.

**Comment by joshtriplett** [comment]
@ehuss
> Are you aware of examples of tools with environment variables that use case-sensitive mixed-case variable names? It seems very unusual to me.

Yes, typically for this exact reason: a fixed prefix followed by arbitrary user-defined data. Off the top of my head, I know autotools does this quite often.

> I think that's a bit of a stretch, this only matters when the programs are in the same Cargo workspace.

1) Consider a busybox-like project that implements many programs as part of the same project.

2) Even if it doesn't cause breakage in a given project, it still leads to confusion and not a small amount of dismay to run into a program on Linux that changes the case of filenames, as people who have run into such issues before know the potential for such breakage.

**Comment by alexcrichton** [comment]
If `env::var("CARGO_THING_the_binary")` works on all platforms, that's fine by me. I don't mind whether it's uppercase or lowercase, but "the naive thing" should work on all platforms. I'm not sure whether that works on Windows or not, myself, though.

**Comment by joshtriplett** [comment]
@alexcrichton 
> If `env::var("CARGO_THING_the_binary")` works on all platforms, that's fine by me.

It should, as far as I can tell.

**Comment by ehuss** [comment]
I have updated so that it doesn't modify the binary name in the environment variable.  I think it should be fine, though it feels a bit unnatural to me.  @joshtriplett does that resolve your concern?

**Comment by joshtriplett** [comment]
Reviewed. This looks great to me, thank you!

**Comment by joshtriplett** [comment]
@rfcbot resolved should-use-same-case-as-Cargo.toml

@rfcbot reviewed

File: src/doc/man/cargo-test.adoc [hunk 1]
```diff
@@ -74,6 +74,14 @@ ignore the `test` flag and will always test the given target.
 Doc tests for libraries may be disabled by setting `doctest = false` for the
 library in the manifest.
 
+Binary targets are automatically built if there is an integration test or
+benchmark. This allows an integration test to execute the binary to exercise
+and test its behavior. The `CARGO_BIN_EXE_<name>`
+linkcargo:reference/environment-variables.html#environment-variables-cargo-sets-for-crates[environment variable]
+is set when the integration test is built so that it can use the
+link:https://doc.rust-lang.org/std/macro.env.html[`env` macro] to locate the
```
Comment (lzutao) [inline src/doc/man/cargo-test.adoc:82]: This seems wrong.
